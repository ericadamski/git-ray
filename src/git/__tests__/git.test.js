const mockProcess = {};

jest.mock('../utils/spawn.js', () =>
  jest.fn().mockImplementation(() => mockOf(mockProcess))
);

const { Observable, of: mockOf } = require('rxjs');
const spawn = require('../utils/spawn.js');
const git = require('../git');

describe('.git', () => {
  const cmd = 'clone';
  const repo = 'git@github.com/ericadamski/git-ray.git';

  beforeEach(() => {
    spawn.mockClear();
  });

  it('should throw an error if no command is given or it is NOT a string', () => {
    expect(() => git()).toThrowErrorMatchingSnapshot();
    expect(() => git([])).toThrowErrorMatchingSnapshot();
    expect(() => git({})).toThrowErrorMatchingSnapshot();
    expect(() => git(123)).toThrowErrorMatchingSnapshot();
    expect(() => git(null)).toThrowErrorMatchingSnapshot();
    expect(() => git(true)).toThrowErrorMatchingSnapshot();
    expect(() => git(undefined)).toThrowErrorMatchingSnapshot();
  });

  it('should return an observable', () => {
    expect(git(cmd)).toBeInstanceOf(Observable);
  });

  it('should call spawn with params', () => {
    return git(cmd)
      .toPromise()
      .then(() => {
        expect(spawn).toHaveBeenCalledTimes(1);
        expect(spawn).toHaveBeenCalledWith('git', [cmd], {
          cwd: process.cwd(),
        });
      });
  });

  it('should accept extra params', () => {
    const params = ['--depth', '1', '--branch', 'feature/new-branch', repo];

    return git(cmd, params)
      .toPromise()
      .then(() => {
        expect(spawn).toHaveBeenCalledTimes(1);
        expect(spawn).toHaveBeenCalledWith('git', [cmd, ...params], {
          cwd: process.cwd(),
        });
      });
  });

  it('should accept options', () => {
    const options = { stdio: 'ignore' };

    return git(cmd, options)
      .toPromise()
      .then(() => {
        expect(spawn).toHaveBeenCalledTimes(1);
        expect(spawn).toHaveBeenCalledWith(
          'git',
          [cmd],
          Object.assign(
            {
              cwd: process.cwd(),
            },
            options
          )
        );
      });
  });

  it('should accept options and params', () => {
    const params = ['--depth', '1', '--branch', 'feature/new-branch', repo];
    const options = { stdio: 'ignore', cwd: '/tmp' };

    return git(cmd, params, options)
      .toPromise()
      .then(() => {
        expect(spawn).toHaveBeenCalledTimes(1);
        expect(spawn).toHaveBeenCalledWith('git', [cmd, ...params], options);
      });
  });
});
