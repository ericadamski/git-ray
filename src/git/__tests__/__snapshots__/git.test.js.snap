// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.git should throw an error if no command is given or it is NOT a string 1`] = `"Refusing to run git. Invalid command undefined."`;

exports[`.git should throw an error if no command is given or it is NOT a string 2`] = `"Refusing to run git. Invalid command ."`;

exports[`.git should throw an error if no command is given or it is NOT a string 3`] = `"Refusing to run git. Invalid command [object Object]."`;

exports[`.git should throw an error if no command is given or it is NOT a string 4`] = `"Refusing to run git. Invalid command 123."`;

exports[`.git should throw an error if no command is given or it is NOT a string 5`] = `"Refusing to run git. Invalid command null."`;

exports[`.git should throw an error if no command is given or it is NOT a string 6`] = `"Refusing to run git. Invalid command true."`;

exports[`.git should throw an error if no command is given or it is NOT a string 7`] = `"Refusing to run git. Invalid command undefined."`;
